internal/export/installedobjects.go

```diff
@@ -0,0 +1,280 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package export
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/elastic/elastic-package/internal/elasticsearch"
+)
+
+const (
+	ComponentTemplatesExportDir = "component_templates"
+	ILMPoliciesExportDir        = "ilm_policies"
+	IndexTemplatesExportDir     = "index_templates"
+	IngestPipelinesExportDir    = "ingest_pipelines"
+)
+
+// InstalledObjectsExporter discovers and exports objects installed in Elasticsearch for a given package.
+type InstalledObjectsExporter struct {
+	packageName string
+	client      *elasticsearch.API
+
+	componentTemplates []ComponentTemplate
+	ilmPolicies        []ILMPolicy
+	indexTemplates     []IndexTemplate
+	ingestPipelines    []IngestPipeline
+}
+
+// NewInstalledObjectsExporter creates an InstalledObjectsExporter for a given package.
+func NewInstalledObjectsExporter(client *elasticsearch.API, packageName string) *InstalledObjectsExporter {
+	return &InstalledObjectsExporter{
+		packageName: packageName,
+		client:      client,
+	}
+}
+
+// ExportAll discovers and exports all known resources as files in the given directory.
+func (e *InstalledObjectsExporter) ExportAll(ctx context.Context, dir string) (count int, err error) {
```

Comment:
We know about:
[index_patterns](https://github.com/elastic/package-spec/blob/main/versions/1/kibana/spec.yml#L59)
[ingest_pipelines](https://github.com/elastic/package-spec/blob/main/versions/1/elasticsearch/spec.yml#L15)
[ilm_policies](https://github.com/elastic/package-spec/blob/main/versions/1/data_stream/spec.yml#L39)

On the other hand, the ingest pipeline can appear in different places (`elasticsearch` or `data_stream/elasticsearch`) and it might be hard to figure out where to dump it.

We don't know know what to do with  "component-templates". Hm...

I'm afraid that I can't give a simple answer on what to do. We could introduce a new command like `elastic-package dump installed-objects` and it will create a new directory with dumped resources. This way we can leave export and import to be "organized" load/store actions, which preserve file structure.

Reply:
Interesting, I didn't know the spec also has supports for custom ILM policies and index patterns, though I see this is barely used. For ingest pipelines we would probably want to export them as YAML to make them useful for package developers.

Adding a `dump` command makes sense to me. So we keep `export` for commands that export things in formats more suited to be included in packages. I will do that.